src/compiler/checker.ts

```diff
@@ -5323,22 +5327,17 @@ namespace ts {
             return undefined;
         }
 
-        // A type reference is considered independent if each type argument is considered independent.
-        function isIndependentTypeReference(node: TypeReferenceNode): boolean {
-            if (node.typeArguments) {
-                for (const typeNode of node.typeArguments) {
-                    if (!isIndependentType(typeNode)) {
-                        return false;
-                    }
-                }
-            }
-            return true;
+        /** A type reference is free of this references if each type argument is free of this references. */
+        function isTypeReferenceFreeOfThisReference(node: TypeReferenceNode): boolean {
+            return !node.typeArguments || node.typeArguments.every(isTypeFreeOfThisReference);
         }
 
-        // A type is considered independent if it the any, string, number, boolean, symbol, or void keyword, a string
-        // literal type, an array with an element type that is considered independent, or a type reference that is
-        // considered independent.
-        function isIndependentType(node: TypeNode): boolean {
+        /**
+         * A type is free of this references if it's the any, string, number, boolean, symbol, or void keyword, a string
+         * literal type, an array with an element type that is free of this references, or a type reference that is
+         * free of this references.
+         */
+        function isTypeFreeOfThisReference(node: TypeNode): boolean {

```

Comment:
I would also negate this.

Reply:
I'll try that in a separate PR. I'm afraid that if I introduce any bugs with further refactoring they'll only show up as performance regressions, and I want to isolate that from any required performance regressions that result from the additional instantiations here.